### DISCUSSION TITLE
external program shows 0.0 temps

### DISCUSSION THREAD

**Discussion by waanito** [artisan-roaster-scope/artisan#1113] [Q&A] [answered]
on macos  have written a python script to interface with a Pico TC08 which returns two temps in the format described in the help.
I have the TC08 with 2 thermocouples attached and when I run the script from the commend line like so:

:->[~/Artisan-Scope/artisan-scope-tc08]$ ./tc08read.sh 
66.6,66.8   

as you can see it returns the two temps in deg F in the requested format.
 When I run artisan (2.8.1) using the script it shows
Device set to /Users/jonny/Artisan-Scope/artisan-scope-tc08/tc08read.sh  Now check serial port settings

There don't seem to be any serial port settings available to set.

When Start is clicked artisan displays "callprogram() received: @line758
ET and BT both show 0.0  so there's some kind of disconnect between my program and artisan.

Any clues on how to debug this?

**Comment by waanito** [comment]
hold onto your hat as this gets complicated:
this all cam about as running the python script fro artisan loses the env variable and the script can no longer find the TC08 framework. compounded by not being able to set env from python (but you can read them)

so right after I posted the above I realized easiest way to test was similar to the help,  to set my script as
`#!/bin/sh`
`echo "237.1,100.4"`

and of course that just works as expected: ET 237.1 & BT 100.4  
and the state changes to "scope monitoring..."

so  next test is via python
`#!/usr/bin/env python3`
`print("69.3,69.5")`

well that works too!

back to the working from cli but not from artisan:
`#!/usr/bin/env python3`
`export DYLD_LIBRARY_PATH=/Library/Frameworks/PicoSDK.framework/Libraries/libusbtc08/`
`python3 ./tc08artisan_read.py`

did it that way to allow setting  DYLD_LIBRARY_PATH which otherwise gets lost maybe related to
https://stackoverflow.com/questions/60126159/how-to-set-ld-library-path-dyld-library-path-on-macos
ok that fails again as now expected

how about reducing my tc08 code to a simple script in that subcall to  `python3 ./tc08artisan_read.py`
`#!/usr/bin/env python3`
`print("237.2,100.5")`

that fails too, yet works on cli
thinking it might be a shell return passing problem I tried the suggestion here
https://stackoverflow.com/questions/23256427/return-a-value-from-shell-script-into-another-shell-script

outer shell script:
`#!/bin/sh`
`export DYLD_LIBRARY_PATH=/Library/Frameworks/PicoSDK.framework/Libraries/libusbtc08/`
`# Call the second script and store it's results in this variable:`
`RETURN_VARIABLE=$(./tc08artisan_read.py)`
`# Display the returned variable from the second script:`
`echo $RETURN_VARIABLE`

with second script simple:
`#!/usr/bin/env python3`
`print("123.4,234.5")`

which works fine from cli but alas still fails from artisan
but...if I change the outer script last line to just echo the exact same numbers like so:
`#!/bin/sh`
`export DYLD_LIBRARY_PATH=/Library/Frameworks/PicoSDK.framework/Libraries/libusbtc08/`
`# Call the second script and store it's results in this variable:`
`RETURN_VARIABLE=$(./tc08artisan_read.py)`
`# Display the returned variable from the second script:`
`#echo $RETURN_VARIABLE`
`echo "123.4,234.5"`

it works. I don't get it. Seems like exactly same thing being returned but clearly to artisan it isn't

**Comment by MAKOMO** [comment]
Could this be related to [Issue #1092](https://github.com/artisan-roaster-scope/artisan/issues/1092)? Did you try the current Continuous Build to see if this fixes the issue?

**Reply by waanito** [reply]
@MAKOMO  maybe,  it certainly is an ENV issue, my solution is (TL;DR) the last few lines of my previous long winded debug reply here.

**Reply by waanito** [reply]
sigh. That was a ghost fix. All that did was show that the python script can talk to artisan using print()
when I add back all the TC08 reading stuff it once again can't find the DYLD_LIBRARY_PATH to the TC08 framework.
Back to debugging.

**Reply by waanito** [reply]
@MAKOMO this is the issue:
https://stackoverflow.com/questions/60126159/how-to-set-ld-library-path-dyld-library-path-on-macos
"Variables LD_LIBRARY_PATH / DYLD_LIBRARY_PATH are not passed to the environment of a child process on macOS if System Integrity Protect (SIP) is enabled."
and as far as I know I can't set an ENV from python

**Reply by waanito** [reply]
Well we lives and we learns, the fix is really in this time, since you can set ENV from python:
`import os`
`os.environ["DYLD_LIBRARY_PATH"] = "/Library/Frameworks/PicoSDK.framework/Libraries/libusbtc08/"`

and now I can read temps from the tc08 YAY!

**Comment by waanito** [comment]
hold onto your hat as this gets complicated:
this all cam about as running the python script fro artisan loses the env variable and the script can no longer find the TC08 framework. compounded by not being able to set env from python (but you can read them)

so right after I posted the above I realized easiest way to test was similar to the help,  to set my script as
`#!/bin/sh`
`echo "237.1,100.4"`

and of course that just works as expected: ET 237.1 & BT 100.4  
and the state changes to "scope monitoring..."

so  next test is via python
`#!/usr/bin/env python3`
`print("69.3,69.5")`

well that works too!

back to the working from cli but not from artisan:
`#!/usr/bin/env python3`
`export DYLD_LIBRARY_PATH=/Library/Frameworks/PicoSDK.framework/Libraries/libusbtc08/`
`python3 ./tc08artisan_read.py`

did it that way to allow setting  DYLD_LIBRARY_PATH which otherwise gets lost maybe related to
https://stackoverflow.com/questions/60126159/how-to-set-ld-library-path-dyld-library-path-on-macos
ok that fails again as now expected

how about reducing my tc08 code to a simple script in that subcall to  `python3 ./tc08artisan_read.py`
`#!/usr/bin/env python3`
`print("237.2,100.5")`

that fails too, yet works on cli
thinking it might be a shell return passing problem I tried the suggestion here
https://stackoverflow.com/questions/23256427/return-a-value-from-shell-script-into-another-shell-script

outer shell script:
`#!/bin/sh`
`export DYLD_LIBRARY_PATH=/Library/Frameworks/PicoSDK.framework/Libraries/libusbtc08/`
`# Call the second script and store it's results in this variable:`
`RETURN_VARIABLE=$(./tc08artisan_read.py)`
`# Display the returned variable from the second script:`
`echo $RETURN_VARIABLE`

with second script simple:
`#!/usr/bin/env python3`
`print("123.4,234.5")`

which works fine from cli but alas still fails from artisan
but...if I change the outer script last line to just echo the exact same numbers like so:
`#!/bin/sh`
`export DYLD_LIBRARY_PATH=/Library/Frameworks/PicoSDK.framework/Libraries/libusbtc08/`
`# Call the second script and store it's results in this variable:`
`RETURN_VARIABLE=$(./tc08artisan_read.py)`
`# Display the returned variable from the second script:`
`#echo $RETURN_VARIABLE`
`echo "123.4,234.5"`

it works. I don't get it. Seems like exactly same thing being returned but clearly to artisan it isn't

OK the fix is in:
i ditched the shell script and went back to using the python script directly but after adding the  DYLD_LIBRARY_PATH above to PATH in my .bash_profile and rebooting.
Now it just works. whew!

**Reply by waanito** [reply]
that working is only that python can print to artisan...the ENV issue remains

**Comment by waanito** [comment, marked as answer]
answer from above repeated here since it seems you can't set a reply as the answer.

I was able to set the missing environment variable from python like so
`import os`
`os.environ["DYLD_LIBRARY_PATH"] = "/Library/Frameworks/PicoSDK.framework/Libraries/libusbtc08/"`

and now I have PICO TC08 working with artisan.  Yay! thanks to the artisan team for creating such a great app.
